File: scripts/generate_sitemap.py
"""Generate public/sitemap.xml (Python fallback when Node is unavailable)."""
from __future__ import annotations

import os
from datetime import datetime, timezone
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
SITE = "https://mtpcode.com"
OUT = ROOT / "public" / "sitemap.xml"
SKIP_DIRS = {"node_modules", "dist", ".git", "public", "assets", "scripts", "legal"}


def collect_html(dir_path: Path, prefix: str = "") -> list[dict]:
    pages: list[dict] = []
    for name in sorted(os.listdir(dir_path)):
        full = dir_path / name
        if full.is_dir():
            if name in SKIP_DIRS:
                continue
            pages.extend(collect_html(full, f"{prefix}{name}/"))
        elif name.endswith(".html") and name != "404.html":
            pages.append({"path": f"{prefix}{name}", "mtime": full.stat().st_mtime})
    return pages


def loc_for(path: str) -> str:
    return f"{SITE}/" if path == "index.html" else f"{SITE}/{path}"


def meta_for(path: str) -> tuple[str, str]:
    if path == "index.html":
        return "weekly", "1.0"
    if path in {"apps.html", "projects.html", "blog.html", "downloads.html"}:
        return "weekly", "0.9"
    if path.startswith("apps/"):
        return "monthly", "0.8"
    if path.startswith("projects/"):
        return "monthly", "0.7"
    if path.startswith("blog/"):
        return "yearly", "0.6"
    if path in {"privacy.html", "terms.html"}:
        return "yearly", "0.3"
    return "monthly", "0.6"


def main() -> None:
    pages = collect_html(ROOT)
    pages.sort(key=lambda p: (p["path"] != "index.html", p["path"]))
    lines = []
    for p in pages:
        changefreq, priority = meta_for(p["path"])
        lastmod = datetime.fromtimestamp(p["mtime"], tz=timezone.utc).strftime("%Y-%m-%d")
        lines.append(
            f"  <url><loc>{loc_for(p['path'])}</loc><lastmod>{lastmod}</lastmod>"
            f"<changefreq>{changefreq}</changefreq><priority>{priority}</priority></url>"
        )
    xml = (
        '<?xml version="1.0" encoding="UTF-8"?>\n'
        '<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">\n'
        + "\n".join(lines)
        + "\n</urlset>\n"
    )
    OUT.write_text(xml, encoding="utf-8")
    print(f"sitemap.xml → {OUT.relative_to(ROOT)} ({len(pages)} URLs)")


if __name__ == "__main__":
    main()

File: scripts/generate-sitemap.js
/**
 * Generates public/sitemap.xml from Vite HTML page entries.
 * Runs on every `npm run build` (via prebuild) so Search Console stays in sync.
 *
 * Excludes: 404.html, /legal/** (robots.txt Disallow), public/, node_modules/, dist/
 */
import { readdirSync, statSync, writeFileSync } from 'node:fs';
import { resolve, relative, dirname } from 'node:path';
import { fileURLToPath } from 'node:url';

const __dirname = dirname(fileURLToPath(import.meta.url));
const root = resolve(__dirname, '..');
const SITE = 'https://mtpcode.com';
const OUT = resolve(root, 'public/sitemap.xml');

const SKIP_DIRS = new Set(['node_modules', 'dist', '.git', 'public', 'assets', 'scripts']);

function collectHtml(dir, prefix = '') {
  const pages = [];
  for (const name of readdirSync(dir)) {
    const full = resolve(dir, name);
    const st = statSync(full);
    if (st.isDirectory()) {
      if (SKIP_DIRS.has(name) || name === 'legal') continue;
      pages.push(...collectHtml(full, `${prefix}${name}/`));
    } else if (name.endsWith('.html') && name !== '404.html') {
      pages.push({
        path: `${prefix}${name}`,
        mtime: st.mtime,
      });
    }
  }
  return pages;
}

function locFor(path) {
  if (path === 'index.html') return `${SITE}/`;
  return `${SITE}/${path}`;
}

function metaFor(path) {
  if (path === 'index.html') return { changefreq: 'weekly', priority: '1.0' };
  if (['apps.html', 'projects.html', 'blog.html', 'downloads.html'].includes(path)) {
    return { changefreq: 'weekly', priority: '0.9' };
  }
  if (path.startsWith('apps/')) return { changefreq: 'monthly', priority: '0.8' };
  if (path.startsWith('projects/')) return { changefreq: 'monthly', priority: '0.7' };
  if (path.startsWith('blog/')) return { changefreq: 'yearly', priority: '0.6' };
  if (path === 'privacy.html' || path === 'terms.html') {
    return { changefreq: 'yearly', priority: '0.3' };
  }
  return { changefreq: 'monthly', priority: '0.6' };
}

function formatDate(d) {
  return d.toISOString().slice(0, 10);
}

const rootPages = collectHtml(root);
rootPages.sort((a, b) => {
  if (a.path === 'index.html') return -1;
  if (b.path === 'index.html') return 1;
  return a.path.localeCompare(b.path);
});

const urls = rootPages
  .map(({ path, mtime }) => {
    const { changefreq, priority } = metaFor(path);
    return `  <url><loc>${locFor(path)}</loc><lastmod>${formatDate(mtime)}</lastmod><changefreq>${changefreq}</changefreq><priority>${priority}</priority></url>`;
  })
  .join('\n');

const xml = `<?xml version="1.0" encoding="UTF-8"?>
<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">
${urls}
</urlset>
`;

writeFileSync(OUT, xml, 'utf8');
console.log(`sitemap.xml → ${relative(root, OUT)} (${rootPages.length} URLs)`);
